### ISSUE TITLE
Support `terraform import` via the cdk cli

### ISSUE THREAD

**Issue by fmmoret** [hashicorp/terraform-cdk#1134]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Support importing existing resources into a stack's state file.

### References

https://www.terraform.io/docs/cli/import/index.html

**Comment by remyleone** [comment]
That would be very useful to have a way to generate HCL configuration from existing configuration such as https://github.com/cycloidio/terracognita seems to provide for different vendors.

**Comment by DanielMSchmidt** [comment]
With the `cdktf convert` command you can take the output of terracognita and use it to create an equivalent CDK program

**Comment by jsteinich** [comment]
@rirze Thanks for sharing your workflow.

I could see having some like `importFrom` available on all resources. This would basically just do what you are already doing with an override and a basic script of Terraform commands. That feels like a relatively low scope way of providing import support.
Could perhaps add support for some custom setting/generation for resources with strange semantics. Likely wouldn't want to go to far down this route though.

Might be interesting to have some sort of `createResource` method available on data sources which effectively does an import, but with the necessary id queried through friendlier properties. This would be harder to automatically generate and doesn't directly fit within current Terraform operations, but could be pretty slick.

**Comment by rirze** [comment]
@jsteinich Glad to see some feedback on this-- I think providing a basic level of "use this resource_id when using `terraform import`" would go a long way for this feature. I think due to the high level of you guys are trying to support, I think it'll be a while before a more native approach would manifest. Even for the peculiar case of `aws_security_group_rule`, I could support this workflow by making API calls during the synthesize process (which I'm not too keen on, but just saying it is possible). 

If you want to move forward with this idea, do you see a particular way of exposing this to the user? I've seen @ansgarm detail an approach using an `Imports.of(resource...).addImportId(resource_id)` pattern (https://github.com/hashicorp/terraform-cdk/pull/1543#issuecomment-1033611825). That seems like the easiest way to implement this, although it might be a little verbose for someone not familiar with that style.

**Comment by jsteinich** [comment]
> If you want to move forward with this idea, do you see a particular way of exposing this to the user? I've seen @ansgarm detail an approach using an `Imports.of(resource...).addImportId(resource_id)` pattern ([#1543 (comment)](https://github.com/hashicorp/terraform-cdk/pull/1543#issuecomment-1033611825)). That seems like the easiest way to implement this, although it might be a little verbose for someone not familiar with that style.

I see that approach as more of just a friendlier way for a user to add additional functionality to a resource (in this case adding some import metadata). Building directly into the api we could add `importFrom` directly onto `TerraformResource`. This could either get synthesized out to a "comment" in the json file that is later acted upon, or be used by a new `import` command variant to generate and execute the native Terraform commands.

**Comment by ansgarm** [comment]
> I see that approach as more of just a friendlier way for a user to add additional functionality to a resource (in this case adding some import metadata)

Yes, that was exactly I was after there. It was my proposal as an alternative to an "add any metadata" method. However, in case of the import use-case we're discussing here, I'd go with an API as proposed by @jsteinich.
And you're both right, we should dare to add such am API rather earlier than later. I was probably a bit too defensive in getting the workflow exactly right first 😅

Does anyone of you know whether there is any documentation on the id that one has to supply to `terraform import`? I remember that it sometimes took the `arn` but not always (for the AWS provider at least). If we had that information _somewhere_ we could use it in docstrings and maybe make that part of the generated provider bindings. But I fear we don't (which is also fine, I think I just always wished these docs existed back when I used `terraform import` 😄).

**Comment by jsteinich** [comment]
> I don't know if this `Import` category is available when `terraform-cdk` builds providers, but I would guess that's the easiest way to generate documentation for this feature.

It's not something that's directly available, but the idea of looking at the docs during generation has been brought up before. Might also be feasible to get it included in the upstream Terraform schema.

**Comment by RichiCoder1** [comment]
The AWS CDK has a surprisingly decent flow that could be emulated here: https://github.com/aws/aws-cdk/tree/main/packages/aws-cdk#cdk-import

TL;DR:
* Add resources to stack
* Run `cdktf import`
* `cdktf` synthesizes and then prompts for any resources that would be added (present in stack but not remote state); This is probably the Hard Part ™️. This could also be a place where the CLI surfaces doc links for imports
* `cdktf` does the necessary state imports for resources

**Comment by rirze** [comment]
> * Add resources to stack
> * Run `cdktf import`
> * `cdktf` synthesizes and then prompts for any resources that would be added (present in stack but not remote state); This is probably the Hard Part ™️. This could also be a place where the CLI surfaces doc links for imports
> * `cdktf` does the necessary state imports for resources

It's a very manual process imo. From what I remember, the CLI prompts for the ARN (AWS specific in this case) one by one. Might be a good initial step but I don't think this should be the long term solution.

**Comment by keidarcy** [comment]
Finally, I found this workaround to import existing resources.

1. assume resources exist, write the approximate cdktf code.
2. run `cdkft synth` in project root.
3. `cd cdktf.out/stacks/<stack>`
4. check `uniqueId` in `cdk.tf.json`, it's located at
```
  "resource": {
    "<Resource_Name>": {
      "<Unique_ID>": {
        "//": {
          "metadata": {
            "path": "<Path>",
            "uniqueId": "<Unique_ID>"
          }
        },
``` 
5. `terraform import <Resource_Name>.<Unique_ID> <ID>`
6. `cd ../../../`, `cdktf diff <stack>`

You should see `change` instead of `create`.

**Comment by DanielMSchmidt** [comment]
We also recently published some docs around refactoring. If you move constructs between stacks you also need to import, maybe it is helpful to anyone reading this: https://developer.hashicorp.com/terraform/cdktf/examples-and-guides/refactoring

**Comment by PookMook** [comment]
> Finally, I found this manual workaround to import existing resources.
> 
> 1. assume resources exist, write the approximate cdktf code.
> 2. run `cdkft synth` in project root.
> 3. `cd cdktf.out/stacks/<stack>`
> 4. check `uniqueId` in `cdk.tf.json`, it's located at
> 
> ```
>   "resource": {
>     "<Resource_Name>": {
>       "<Unique_ID>": {
>         "//": {
>           "metadata": {
>             "path": "<Path>",
>             "uniqueId": "<Unique_ID>"
>           }
>         },
> ```
> 
> 5. `terraform init` and `terraform import <Resource_Name>.<Unique_ID> <ID>`
> 6. `cd ../../../`, `cdktf diff <stack>`
> 
> You should see `change` instead of `add`.

Thanks for that message, works like a charm. Make sure you use the "DB instance ID" for <ID> on the DB you want to import, not the ARN. I guess it works the same for other types of ressources. Makes life easier and prevent the string too long errors

**Comment by Danielku15** [comment]
The workaround from @keidarcy works indeed nicely. We are using the CDK with C# and I wrote some code which automates this for us until there is an official feature. 

1. We run the synth command
2. We then read all resources listed in the resulting cdk.tf.json 
3. We read also the existing tfstate and remove all resources which are already tracked in the state. 
4. We run a `terraform init`
5. We then try to run a `terraform import` for all resources which are not known in the state. 

Of course there can be errors in cases where you mix actually new items and items to import in one run but we can deal with that. The biggest challenge was building the import IDs which depend on the provider and resource you are importing and you need to align this with the docs. 

As we are doing a larger migration it was worth automating this for us.

**Comment by mackignacio** [comment]
Hi everyone. I created a CLI for [CDKTF import](https://www.npmjs.com/package/cdktf-import) using the solution provided from this [comment](https://github.com/hashicorp/terraform-cdk/issues/1134#issuecomment-1396684489) of @keidarcy and @Danielku15 . This is still on development but this is better than doing this manually.

**Comment by mackignacio** [comment]
Nice Thank you. For a moment Im just using mine.
